BIP Draft: Standard Encrypted Wallet Payload

KeysSoze opened the pull request:
This pull request proposes a new BIP for a standard, interoperable format for encrypted wallet backups and transfers.

The core design choice is the separation of the data `payload` from the cryptographic `container`. This BIP specifies the `payload` format, while mandating that it must be encrypted by a container defined in companion proposals.

Key features include:
* **CBOR Serialisation:** Uses CBOR for a compact, efficient, and extensible binary format suitable for resource-constrained devices.
* **Descriptor-Centric:** Aligns with modern wallet architecture by using BIP 380 Output Script Descriptors as the primary method for defining accounts.
* **Comprehensive Data:** The schema includes accounts, master secret material, transaction data, UTXOs and standardised metadata.

This proposal was previously posted to the Bitcoin-Dev mailing list: [https://groups.google.com/g/bitcoindev/c/3I-qtBhzIGY](https://groups.google.com/g/bitcoindev/c/3I-qtBhzIGY)

Seeking assignment of a BIP number and welcoming all feedback.

KeysSoze committed:
Standard Encrypted Wallet Payload

murchandamus reviewed (commented):
Hello @KeysSoze,

This document is incorrectly formatted in a manner that makes me think that it was AI generated. If you are still interested in pursuing this proposal, could you please fix the formatting?

At this time, given the lack of replies on the mailing list and comments on this PR, it is not clear to me that this proposal is being pursued for implementation by any Bitcoin project. Given that there are several other backup related proposals in flight, perhaps some of the corresponding authors would be interested in reviewing your document if you review theirs.

I‘d be happy to provide a more thorough review when this proposal has gotten review from other parties and that review has been addressed, or the proposal otherwise distinguishes itself from its perceived AI-generated origin.

KeysSoze committed:
Updated formatting

Changed preamble to more closely follow BIP standards
Removed table of contents

KeysSoze committed:
Remove superfluous html `<span>` tags

KeysSoze commented:
Hi @murchandamus,

> here are some similar proposals that have been recently merged or are still in flight
> 
> * [BIP139: Wallet Metadata Backup Format](https://github.com/bitcoin/bips/pull/2130)
> 
> which might interest you.


Yes, it seems BIP 139 is quite similar to this proposal. But rather than using JSON, this proposal defines a rigid CDDL schema with deterministic CBOR, enforcing true inter-wallet compatibility. CBOR's compact binary nature drastically reduces the payload size making air-gapped QR codes or NFC transfers highly viable if defined in a companion BIP. Furthermore, the modular nature of this proposal delegates security to battle-tested IETF standards like COSE via a separate companion BIP. This ensures the standard is inherently future-proofed. As cryptography advances developers can seamlessly upgrade ciphers, implement multi-party authorisation or adopt post-quantum algorithms within the security container without ever needing to alter the core wallet payload format.

murchandamus commented on bip-wallet-payload.mediawiki:
```diff
@@ -0,0 +1,791 @@
+  Title: Standard Encrypted Wallet Payload
+  Authors: Keyser Söze <keys.soze@proton.me>
+  Status: Draft
+  Type: Standards Track
```
```suggestion
  Type: Specification
```

murchandamus commented on bip-wallet-payload.mediawiki:
```diff
@@ -0,0 +1,791 @@
+|-
+| 3 || root || map || The wallet’s master secret material. || No
+|-
+| 10 || accounts || [+ map] || An array of account maps, each defined by a descriptor. || Yes
```
Above it says that an account references one or more descriptors, but here it says that the accounts are defined by descriptors. This makes it sound like the descriptors are both keys and values.

murchandamus commented on bip-wallet-payload.mediawiki:
```diff
@@ -0,0 +1,791 @@
+|-
+| 0 || mainnet || The main Bitcoin production network.
+|-
+| 1 || testnet || The primary Bitcoin test network.
```
There are multiple active testnets. If they are meant to be disambiguated per type + genesis_hash that could be communicated more explicitly.

murchandamus commented on bip-wallet-payload.mediawiki:
```diff
@@ -0,0 +1,791 @@
+|-
+| 1 || testnet || The primary Bitcoin test network.
+|-
+| 2 || signet || The signet test network.
```
While there is one default signet, there are also popular other signets (such as e.g., MutinyNet) and they are by design easy to create for anyone.

murchandamus commented on bip-wallet-payload.mediawiki:
```diff
@@ -0,0 +1,791 @@
+The root map contains the wallet’s master secret material from which all accounts and descriptors are derived. It may include the BIP-39 mnemonic, optional passphrase, raw entropy, or the derived BIP-39 seed. The root map is optional and a wallet '''MUST NOT''' contain more than one.
+
+{| class="wikitable"
+! Integer Key !! String Key !! Type !! Description !!
```
```suggestion
! Integer Key !! String Key !! Type !! Description !! Required
```

Is this missing Required?

murchandamus commented on bip-wallet-payload.mediawiki:
```diff
@@ -0,0 +1,791 @@
+network-type = 0 / 1 / 2 / 3  ; 0=mainnet,1=testnet,2=signet,3=regtest
+role-type    = 0 / 1          ; 0=receive,1=change
+txid-type = bstr .size 32
+address-type = p2pkh-address / p2sh-address / segwit-address
```
How about taproot?

murchandamus reviewed (commented):
I gave this a first read. Has any Bitcoin project adopted this scheme or is working on adopting it? Please add the required Backwards Compatibility section.
